plugins/node/opentelemetry-instrumentation-knex/src/instrumentation.ts

```diff
@@ -0,0 +1,225 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as api from '@opentelemetry/api';
+import { VERSION } from './version';
+import * as constants from './constants';
+import {
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationNodeModuleFile,
+  isWrapped,
+} from '@opentelemetry/instrumentation';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import * as utils from './utils';
+import * as types from './types';
+
+import type * as knex from 'knex';
+
+const contextSymbol = Symbol('opentelemetry.instrumentation-knex.context');
+const DEFAULT_CONFIG: types.KnexInstrumentationConfig = {
+  maxQueryLength: 1022,
+};
+
+export class KnexInstrumentation extends InstrumentationBase<typeof knex> {
+  constructor(config: types.KnexInstrumentationConfig = {}) {
+    super(
+      `@opentelemetry/instrumentation-${constants.MODULE_NAME}`,
+      VERSION,
+      Object.assign({}, DEFAULT_CONFIG, config)
+    );
+  }
+
+  init() {
+    const module = new InstrumentationNodeModuleDefinition<typeof knex>(
+      constants.MODULE_NAME,
+      constants.SUPPORTED_VERSIONS
+    );
+
+    module.files.push(
+      this.getClientNodeModuleFileInstrumentation('src'),
+      this.getClientNodeModuleFileInstrumentation('lib'),
+      this.getRunnerNodeModuleFileInstrumentation('src'),
+      this.getRunnerNodeModuleFileInstrumentation('lib'),
+      this.getRunnerNodeModuleFileInstrumentation('lib/execution')
+    );
+
+    return module;
+  }
+
+  private getRunnerNodeModuleFileInstrumentation(basePath: string) {
+    return new InstrumentationNodeModuleFile<typeof knex>(
+      `knex/${basePath}/runner.js`,
+      constants.SUPPORTED_VERSIONS,
+      (Runner: any, moduleVersion) => {
+        api.diag.debug(
+          `Applying runner.js patch for ${constants.MODULE_NAME}@${moduleVersion}`
+        );
+        this.ensureWrapped(
+          moduleVersion,
+          Runner.prototype,
+          'query',
+          this.createQueryWrapper(moduleVersion)
+        );
+        return Runner;
+      },
+      (Runner: any, moduleVersion) => {
+        api.diag.debug(
+          `Removing patch for ${constants.MODULE_NAME}@${moduleVersion}`
+        );
+        this._unwrap(Runner.prototype, 'query');
+        return Runner;
+      }
+    );
+  }
+
+  private getClientNodeModuleFileInstrumentation(basePath: string) {
+    return new InstrumentationNodeModuleFile<typeof knex>(
+      `knex/${basePath}/client.js`,
+      constants.SUPPORTED_VERSIONS,
```

Comment:
Now that I look at it again, maybe we can specify just the relevant versions here for each `basePath`?
e.g. - pass `supportedVersions` parameter to the `getClientNodeModuleFileInstrumentation` function, so that each `basePath` will try to patch only the relevant versions for it.
Currenly, SUPPORTED_VERSIONS is documenting the various permutations, but we will attempt to patch all of them for all versions:
```js
export const SUPPORTED_VERSIONS = [
  // use "lib/execution" as basepath, latest tested 0.95.6
  '>=0.22.0',
  // use "lib" as basepath
  '>=0.10.0 <0.18.0',
  '>=0.19.0 <=0.20.10',
  '>=0.20.11 <0.22.0',
  // use "src" as basepath
  '>=0.18.0 <0.19.0',
];
```

Not required, as the current approach works, but will be more precise and readable I think.